airflow/providers/google/cloud/operators/bigquery.py

```diff
@@ -1692,32 +1693,48 @@ def prepare_template(self) -> None:
             with open(self.configuration, 'r') as file:
                 self.configuration = json.loads(file.read())
 
+    def _submit_job(self, hook: BigQueryHook, job_id: str):
+        # Submit a new job
+        job = hook.insert_job(
+            configuration=self.configuration,
+            project_id=self.project_id,
+            location=self.location,
+            job_id=job_id,
+        )
+        # Start the job and wait for it to complete and get the result.
+        job.result()
+        return job
+
     def execute(self, context: Any):
         hook = BigQueryHook(
             gcp_conn_id=self.gcp_conn_id,
             delegate_to=self.delegate_to,
         )
 
-        job_id = self.job_id or f"airflow_{self.task_id}_{int(time())}"
+        exec_date = re.sub("\:|\-|\+", "_", context['execution_date'].isoformat())
+        job_id = self.job_id or f"airflow_{self.dag_id}_{self.task_id}_{exec_date}_"
+
         try:
-            job = hook.insert_job(
-                configuration=self.configuration,
-                project_id=self.project_id,
-                location=self.location,
-                job_id=job_id,
-            )
-            # Start the job and wait for it to complete and get the result.
-            job.result()
+            # Submit a new job
+            job = self._submit_job(hook, job_id)
         except Conflict:
+            # If the job already exists retrieve it
```

Comment:
what does this do to the the behavior for re-running a DAG for an execution date? 
Would this not by default reattach to the originally succeeded job (rather than expected behavior or re-running the job)? Should this be configurable?

Reply:
My question is: if job succeded why should we rerun it instead of using the existing result?

Reply:
In short, the underlying table / partition could have changed between the original run and the new run and in many cases the user will want to ensure that the result of this job is based on a fresh run of the query.

Airflow cannot know if the underlying table(s) have changed since the original run.

To give an example use case:
Imagine a streaming job in charge of inserting records to an hourly partitioned fact table and a DAG responsible for running some hourly analytics query (joins to dimension tables / aggregation). Once the processing time watermark passes the end of the partition interval original scheduled dag run runs (BigQueryInsertJobOperator task runs a query on the data at this processing time). Then late data (event time << processing time) arrives that should still go to the old partition (because partitioning is on event time). When the streaming job detects this it could use some mechanism (e.g. the REST API or an alert to a human who manually re-runs the dag for an execution date) to tell the DAG (and this BigQueryInsertJobOperator) to re-run because the original run for this interval / partition was actually against incomplete data.

Reply:
@turbaszek I tend to agree with @jaketf.  We frequently might want to rerun a DAG for a prior, successful execution date at some point in the future that would overwrite data in a partitioned BigQuery table.  Typically this might occur because of a change in the business logic.

So, in my opinion, the `job_id` needs to be more unique than just `airflow_{self.dag_id}_{self.task_id}_{exec_date}_` in order to account for this.

If I clear the status of a task in a DAG, I expect that task to be fully rerun.

Reply:
I think we will not be able to create universal behavior that will keep idempodentency. Some users want to always execute SQL query, others want to use predetermined results, and others want to re-query only when it is backfill. All the cases sound correct and they cannot always be combined into a single operator.

Reply:
Then there should be additional parameters.  IMHO, I don't think it is controversial to suggest that, when the status of a BQ task is cleared, the default behaviour should be that the query if fully rerun.  Other cases sound like less common scenarios (in my experience at least).

Reply:
> the query if fully rerun.

That was the original behaviour as the `job_id` was always generated by discovery API.

Reply:
Indeed.  I'd say that should be the default going forward too.

Reply:
> I think we will not be able to create universal behavior that will keep idempotency. Some users want to always execute SQL query, others want to use predetermined results, and others want to re-query only when it is backfill. All the cases sound correct and they cannot always be combined into a single operator without additional parameters.

This is an interesting point. Perhaps we consider bringing back the deprecated BigQueryOperator (contract is always runs a new query) and than have this BigQueryInsertJobOperator (contract is create job if not succeeded exists).

IMHO we can have both behaviors in a single operator with a `force_rerun=True` parameter that controls how the job id is generated.
```python3
# choice of uuid 
uniqueness_hash = hash(uuid.uuid4()) if self.force_rerun else hash(job_config)

job_id = f"{dag_id}{task_id}{exec_date}{uniqueness_hash}"
```